docs: multimodal (#777)
langchain-ai/langsmith-docs: docs/evaluation/how_to_guides/evaluate_with_attachments.mdx

  PythonBlock,
  TypeScriptBlock,
} from "@site/src/components/InstructionsWithCode";
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Run an evaluation with multimodal content

> LangSmith supports creating dataset examples with file attachments, which you can consume when running evals over that dataset.

Reviewer: ```suggestion
LangSmith lets you create dataset examples with file attachments—like images, audio files, or documents—so you can reference them when evaluating an application that uses multimodal inputs or outputs.
```

> Attachments are particularly valuable for LLM applications that work with multimodal inputs or outputs, such as images, audio, or documents. While you can include multimodal data in your examples by base64 encoding it, this approach is inefficient - the encoded data takes up more space than the original binary files, resulting in slower transfers to and from LangSmith. Using attachments instead provides two key benefits:

Reviewer: ```suggestion
While you can include multimodal data in your examples by base64 encoding it, this approach is inefficient - the encoded data takes up more space than the original binary files, resulting in slower transfers to and from LangSmith. Using attachments instead provides two key benefits:
```

1. Faster upload and download speeds due to more efficient binary file transfers
2. Enhanced visualization of different file types in the LangSmith UI

<Tabs className="interface-tabs" queryString="mode">
  <TabItem value="sdk" label="SDK">

> ## Create examples with attachments

Reviewer: ```suggestion
## 1. Create examples with attachments
```

:::

> ## Run evaluations

Reviewer: ```suggestion
## 2. Run evaluations
```

groupId="client-language"
/>

  </TabItem>
  <TabItem value="ui" label="UI">

> ## Create examples with attachments

Reviewer: ```suggestion
## 1. Create examples with attachments
```

You can add examples with attachments to a dataset in a few different ways.

### From existing runs

When adding runs to a LangSmith dataset, attachments can be selectively propagated from the source run to the destination example.
To learn more, please see [this guide](./manage_datasets_in_application#add-runs-from-the-tracing-project-ui).

![](./static/add_trace_with_attachments_to_dataset.png)

### From scratch

> You can upload examples with attachments directly from the LangSmith UI. Click the `+ Example` button in the `Examples` tab of the dataset UI.

Reviewer: ```suggestion
You can create examples with attachments directly from the LangSmith UI. Click the `+ Example` button in the `Examples` tab of the dataset UI.
```

Then upload attachments using the "Upload Files" button:

![](./static/create_example_with_attachments.png)

Once uploaded, you can view examples with attachments in the LangSmith UI. Each attachment will be rendered with a preview for easy inspection.
![](./static/attachments_with_examples.png)

> ## Create a multimodal prompt

Reviewer: ```suggestion
## 2. Create a multimodal prompt
```

The LangSmith UI allows you to include attachments in your prompts when evaluating multimodal models:

First, click the file icon in the message where you want to add multimodal content.
Next, add a template variable for the attachment(s) you want to include for each example.
- For a single attachment type: Use the suggested variable name. Note: all examples must have an attachment with this name.
- For multiple attachments or if your attachments have varying names from one example to another: Use the `All attachments` variable to include all available attachments for each example.

![](./static/adding_multimodal_variable.gif)

## Update examples with attachments

:::note Attachment Size Limit
Attachments are limited to 20MB in size in the UI.
:::

When editing an example in the UI, you can:
- Upload new attachments
- Rename and delete attachments  
- Reset attachments to their previous state using the quick reset button

Changes are not saved until you click submit.

![](./static/attachment_editing.gif)

> ## Define custom evaluators

Reviewer: I'd move this section up above the update section and follow the same section header as the SDK. So have Run evaluations as the header and then Create a multimodal prompt, define custom evaluators underneath

would also help to give an examples of what IS possible. Eg: 

Even without multimodal support in your evaluators, you can still run text‐only evaluations. For example:
- OCR → text correction: Use a vision model to extract text from a document, then evaluate the accuracy of the extracted output.
- Speech‑to‑text → transcription quality: Use a voice model to transcribe audio to text, then evaluate the transcription against your reference.

Author: nice, done
